subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AnnotationProcessorDetectionIntegrationTest.groovy

```diff
@@ -0,0 +1,188 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.java.compile.incremental
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.CompilationOutputsFixture
+import org.gradle.language.fixtures.IncrementalAnnotationProcessorFixture
+import org.gradle.language.fixtures.NonIncrementalAnnotationProcessorFixture
+
+class AnnotationProcessorDetectionIntegrationTest extends AbstractIntegrationSpec {
+
+    CompilationOutputsFixture outputs
+
+    def setup() {
+        outputs = new CompilationOutputsFixture(file("app/build/classes"))
+        createBothProcessors()
+    }
+
+    def "build should be incremental when all APs are incremental"() {
+        appWithIncapProcessorOnly()
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        succeeds 'app:compileJava'
+
+        then:
+        file('app/build/generated-sources/AppIncremental.java').exists()
+        !file('app/build/generated-sources/AppNonIncremental.java').exists()
+
+        when:
+        // Modify main source file.  Would normally be incremental.
+        file('app/src/main/java/App.java').text = "@Incremental class App { /*change*/ }"
+        succeeds 'app:compileJava'
+
+        then:
+        skipped ':inc:compileJava'
+
+        // Verify that it was incremental (B not recompiled).
+        outputs.recompiledClasses 'App', 'AppIncremental'
+    }
+
+    def "presence of non-incremental AP should force non-incremental build"() {
+        appWithBothProcessors()
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        succeeds 'app:compileJava'
+
+        then:
+        file('app/build/generated-sources/AppIncremental.java').exists()
+        file('app/build/generated-sources/AppNonIncremental.java').exists()
+
+        when:
+        // Trivial reformatting of the main source file.  Would normally be incremental.
+        file('app/src/main/java/App.java').text = "@Incremental @NonIncremental class App { /*change*/ }"
+        succeeds 'app:compileJava'
+
+        then:
+        skipped ':inc:compileJava'
+        skipped ':noninc:compileJava'
+
+        // Verify app was rebuilt non-incrementally.
+        outputs.recompiledClasses 'App', 'B', 'AppNonIncremental', 'AppIncremental'
+    }
+
+    def "deleting an annotation should delete its generated file"() {
+        appWithBothProcessors()
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        file('app/src/main/java/App.java').text = "@Incremental class App { /*removed annotation*/ }"
+        succeeds 'app:compileJava'
+
+        then:
+        outputs.deletedClasses 'AppNonIncremental'
+    }
```

Comment:
please add a test where you start with 
```java
@Incremental 
class  App {
}
````
and then remove the annotation. Unfortunately, it will be removed because of the bug.

An even better would be to generate a file that doesn't have the prefix `App`, and then the deletion would prove M1 works and it's not the bug of gradle that deletes it.

Reply:
Which bug are you referring to? We only delete classes starting with `App$`. `AppIncremental` and `AppNonIncremental` do not match that pattern.

Reply:
I've added a test in the next PR update that does this, and shows that there is indeed no bug.